pyccl/tracers.py

```diff
@@ -0,0 +1,405 @@
+                input arrays.
+        """
+        is_factorizable = transfer_ka is None
+        is_k_constant = (transfer_ka is None) and (transfer_k is None)
+        is_a_constant = (transfer_ka is None) and (transfer_a is None)
+        is_kernel_constant = kernel is None
+
+        chi_s, wchi_s = _check_array_params(kernel)
+        if is_factorizable:
+            a_s, ta_s = _check_array_params(transfer_a)
+            lk_s, tk_s = _check_array_params(transfer_k)
+            tka_s = NoneArr
+            if (not is_a_constant) and (a_s.shape != ta_s.shape):
+                raise ValueError("Time-dependent transfer arrays "
+                                 "should have the same shape")
+            if (not is_k_constant) and (lk_s.shape != tk_s.shape):
+                raise ValueError("Scale-dependent transfer arrays "
+                                 "should have the same shape")
+        else:
+            a_s, lk_s, tka_s = _check_array_params(transfer_ka, arr3=True)
+            if tka_s.shape != (len(a_s), len(lk_s)):
+                raise ValueError("2D transfer array has inconsistent "
+                                 "shape. Should be (na,nk)")
+            tka_s = tka_s.flatten()
+            ta_s = NoneArr
+            tk_s = NoneArr
+
+        status = 0
+        ret = lib.cl_tracer_t_new_wrapper(cosmo.cosmo,
+                                          int(der_bessel),
+                                          int(der_angles),
+                                          chi_s, wchi_s,
+                                          a_s, lk_s,
+                                          tka_s, tk_s, ta_s,
+                                          int(is_logt),
+                                          int(is_factorizable),
+                                          int(is_k_constant),
+                                          int(is_a_constant),
+                                          int(is_kernel_constant),
+                                          int(extrap_order_lok),
+                                          int(extrap_order_hik),
+                                          status)
+        self.trc.append(check_returned_tracer(ret))
+
+    def __del__(self):
+        if hasattr(self, 'trc'):
+            for t in self.trc:
+                lib.cl_tracer_t_free(t)
+
+
+class NumberCountsTracer(Tracer):
+    """Specific `Tracer` associated to galaxy clustering with linear
+    scale-independent bias, including redshift-space distortions and
+    magnification.
+
+    Args:
+        cosmo (:obj:`Cosmology`): Cosmology object.
+        has_rsd (bool): Flag for whether the tracer has a
+            redshift-space distortion term.
+        dndz (tuple of arrays): A tuple of arrays (z, N(z))
+            giving the redshift distribution of the objects. The units are
+            arbitrary; N(z) will be normalized to unity.
+        bias (tuple of arrays): A tuple of arrays (z, b(z))
```

Comment:
I'm wondering about the API here. Is there a reason to not require the bias be given?

Reply:
Right now, if bias is `None`, we do not include the density contribution. This can be useful if, for example, you want a magnification-only tracer.